File: pyBASCloudAPI/tests/property_tests.py
import unittest
import time 
import datetime

from definitions import *

import pyBAScloudAPI as api

BCAPI = api.EntityContext(BASCLOUD_TEST_URL)


def errorHandler(e, j):
    raise e

def authenticate():
    currentDateTime = int(time.time())
    print("CURRENT DATETIME: ", currentDateTime)

    currentDateTimeString = datetime.datetime.fromtimestamp(currentDateTime).strftime("%FT%T.000Z")
    print("CURRENT DATETIME STR: ", currentDateTimeString)

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS)

    print("TOKEN: ", BCAPI.getToken())
    print("EXPIRATION DATE: ", datetime.datetime.fromtimestamp(BCAPI.getTokenExpirationDate()).strftime("%FT%T.000Z"))


class TestPropertyCollection(unittest.TestCase):

    def test_property(self):
        print("\tRequesting all properties...")

        props = BCAPI.getPropertiesCollection(BASCLOUD_TEST_TENANT_UUID, errorHandler=errorHandler)
        print("\t\tOK.")

        print("\tFound properties: ", len(props[0]))

        self.assertTrue(len(props[0]) >= 1)


class TestSingleProperty(unittest.TestCase):

    def test_property(self):
        print("\tRequesting single property with UUID...")

        property = BCAPI.getProperty(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_PROPERTY_UUID)
        print("\t\tOK.")

        print("\tProperty UUID: ", property.uuid)
        print("\tProperty Name: ", property.name)
        print("\tProperty Address: {}, {} {}, {}".format(property.street, property.postalCode, property.city, property.country))
        print("\ttProperty created at: ", datetime.datetime.fromtimestamp(property.createdAt).strftime("%FT%T.000Z"))
        print("\ttProperty updated at: ", datetime.datetime.fromtimestamp(property.updatedAt).strftime("%FT%T.000Z"))

        self.assertEqual(property.uuid, BASCLOUD_TEST_PROPERTY_UUID)
        self.assertEqual(property.name, "Nils Test-Property")
        self.assertEqual(property.street, "Dieselstraße 6")
        self.assertEqual(property.postalCode, "63739")
        self.assertEqual(property.city, "Aschaffenburg")
        self.assertEqual(property.country, "Deutschland")
        self.assertTrue(property.createdAt > 0)
        self.assertTrue(property.updatedAt > 0)


class TestAssociatedConnector(unittest.TestCase):

    def test_property(self):
        print("\tRequesting associated connectors of the property...")

        prop_connectors = BCAPI.getAssociatedConnectors(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_PROPERTY_UUID, errorHandler=errorHandler)
        print("\t\tOK.")

        print("\tFound property connectors: ", len(prop_connectors[0]))

        self.assertTrue(len(prop_connectors[0]) >= 1)


class TestCreateUpdateAndDeleteProperty(unittest.TestCase):

    def test_property(self):
        print("\tCreating new property...")

        new_property = BCAPI.createProperty(BASCLOUD_TEST_TENANT_UUID, "TestProperty", "Street", "12345", "City", "Deutschland")
        print("\t\tOK.")

        new_uuid = new_property.uuid

        self.assertEqual(new_property.name, "TestProperty")
        self.assertEqual(new_property.street, "Street")
        self.assertEqual(new_property.postalCode, "12345")
        self.assertEqual(new_property.city, "City")
        self.assertEqual(new_property.country, "Deutschland")
        self.assertTrue(new_property.createdAt > 0)
        self.assertTrue(new_property.updatedAt > 0)

        print("\tUpdating newly created property...")

        up_re_property = BCAPI.updateProperty(BASCLOUD_TEST_TENANT_UUID, new_uuid, "TestProperty 2")
        print("\t\tOK.")

        self.assertEqual(up_re_property.name, "TestProperty 2")

        print("\tDeleting newly created property...")
        BCAPI.deleteProperty(BASCLOUD_TEST_TENANT_UUID, new_uuid)
        print("\t\tOK.")

        try:
            non_property = BCAPI.getProperty(BASCLOUD_TEST_TENANT_UUID, new_uuid)
        except api.NotFoundRequest:
            self.assertTrue(True)
        except:
            self.assertTrue(False)



if __name__ == '__main__':
    authenticate()
    unittest.main()

File: tests/property_tests.cpp
#include <chrono>
#include <iostream>

#include <fmt/core.h>
#include <fmt/chrono.h>

#include <gtest/gtest.h>

#include "EntityContext.h"
#include "entity/EntityCollection.h"
#include "error/Exceptions.h"


using namespace BAScloud;

EntityContext BCAPI(BASCLOUD_TEST_URL);


void authenticate() {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    std::cerr <<"CURRENT DATETIME: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(currentDateTime))<<std::endl;

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS);

    std::cerr <<"TOKEN: "<<BCAPI.getToken()<<std::endl;
    std::cerr <<"EXPIRATION DATE: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(BCAPI.getTokenExpirationDate()))<<std::endl;
}

TEST(BasicTests, PropertyCollectionTest) {

    std::cout << "\tRequesting all properties..." << std::endl;

    EntityCollection<Property> props = BCAPI.getPropertiesCollection(BASCLOUD_TEST_TENANT_UUID, {}, {}, {}, {}, {}, {}, [](std::exception& e, json& j) {
            throw e;
        });
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound properties: " << props.first.size() << std::endl;

    EXPECT_TRUE(props.first.size() >= 1);
}

TEST(BasicTests, SinglePropertyTest) {

    std::cout << "\tRequesting single property with UUID..." << std::endl;

    Property property = BCAPI.getProperty(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_PROPERTY_UUID);
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tProperty UUID: " << property.getUUID() << std::endl;
    std::cout << "\tProperty Name: " << property.getName() << std::endl;
    std::cout << "\tProperty Address: " << fmt::format("{}, {} {}, {}", property.getStreet(), property.getPostalCode(), property.getCity(), property.getCountry()) << std::endl;
    std::cout << "\ttProperty created at: " << fmt::format("{:%F %T}", fmt::localtime(property.getCreatedDate())) << std::endl;
    std::cout << "\ttProperty updated at: " << fmt::format("{:%F %T}", fmt::localtime(property.getLastUpdatedDate())) << std::endl;

    EXPECT_EQ(property.getUUID(), BASCLOUD_TEST_PROPERTY_UUID);
    EXPECT_EQ(property.getName(), "Nils Test-Property");
    EXPECT_EQ(property.getStreet(), "Dieselstraße 6");
    EXPECT_EQ(property.getPostalCode(), "63739");
    EXPECT_EQ(property.getCity(), "Aschaffenburg");
    EXPECT_EQ(property.getCountry(), "Deutschland");
    EXPECT_TRUE(property.getCreatedDate() > 0);
    EXPECT_TRUE(property.getLastUpdatedDate() > 0);
}

TEST(BasicTests, AssociatedConnectorTest) {

    std::cout << "\tRequesting associated connectors of the property..." << std::endl;

    EntityCollection<Connector> prop_connectors = BCAPI.getAssociatedConnectors(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_PROPERTY_UUID, {}, [](std::exception& e, json& j) {
            throw e;
        });
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound property connectors: " << prop_connectors.first.size() << std::endl;

    EXPECT_TRUE(prop_connectors.first.size() >= 1);
}

TEST(BasicTests, CreateUpdateAndDeletePropertyTest) {

    std::cout << "\tCreating new property..." << std::endl;

    Property new_property = BCAPI.createProperty(BASCLOUD_TEST_TENANT_UUID, "TestProperty", "Street", "12345", "City", "Deutschland");
    std::cout << "\t\tOK." << std::endl;

    std::string new_uuid = new_property.getUUID();

    EXPECT_EQ(new_property.getName(), "TestProperty");
    EXPECT_EQ(new_property.getStreet(), "Street");
    EXPECT_EQ(new_property.getPostalCode(), "12345");
    EXPECT_EQ(new_property.getCity(), "City");
    EXPECT_EQ(new_property.getCountry(), "Deutschland");
    EXPECT_TRUE(new_property.getCreatedDate() > 0);
    EXPECT_TRUE(new_property.getLastUpdatedDate() > 0);

    std::cout << "\tUpdating newly created property..." << std::endl;

    Property up_re_property = BCAPI.updateProperty(BASCLOUD_TEST_TENANT_UUID, new_uuid, "TestProperty 2");
    std::cout << "\t\tOK." << std::endl;

    EXPECT_EQ(up_re_property.getName(), "TestProperty 2");

    std::cout << "\tDeleting newly created property..." << std::endl;
    BCAPI.deleteProperty(BASCLOUD_TEST_TENANT_UUID, new_uuid);
    std::cout << "\t\tOK." << std::endl;

    try {
        Property non_property = BCAPI.getProperty(BASCLOUD_TEST_TENANT_UUID, new_uuid);
    } catch(NotFoundRequest& e) {
        EXPECT_TRUE(true);
    } catch(...) {
        EXPECT_TRUE(false);
    }

}


int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);

    authenticate();

    return RUN_ALL_TESTS();
}